### ISSUE TITLE
coq with nix

### ISSUE THREAD

**Issue by viniciusmuller** [ms-jpq/coq_nvim#20]
Basically the same issue as https://github.com/ms-jpq/chadtree/issues/110 (which the provided solution was not working for me some minutes ago).
Nix support would be nice :)

**Comment by ms-jpq** [comment]
wait chadtree broke for you? sorry  can you please elaborate on what broke? thanks!

and yeah I will add that thing in, stay tuned

**Comment by viniciusmuller** [comment]
> wait chadtree broke for you? sorry can you please elaborate on what broke? thanks!

Yes, even though I have `let g:chadtree_settings = {'xdg': v:true}` set up, it still gives the same error that it tried to write in nix's read-only filesystem.

Error message:
```
Error detected while processing function CHADon_stderr:
line    1:
Traceback (most recent call last):
  File "/nix/store/81lwy2hfqj4c1943b1x8a0qsivjhdhw9-python3-3.9.6/lib/python3.9/runpy.py", line 197, in _run_modu
le_as_main
Press ENTER or type command to continue
Error detected while processing function CHADon_stderr:
line    1:
    return _run_code(code, main_globals, None,
  File "/nix/store/81lwy2hfqj4c1943b1x8a0qsivjhdhw9-python3-3.9.6/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/nix/store/p3j4334jxdrdwfqj02zy605jc5k8h16z-vimplugin-chadtree-2021-08-10/share/vim-plugins/chadtree/chad
tree/__main__.py", line 58, in <module>
    EnvBuilder(
  File "/nix/store/81lwy2hfqj4c1943b1x8a0qsivjhdhw9-python3-3.9.6/lib/python3.9/venv/__init__.py", line 67, in cr
eate
    context = self.ensure_directories(env_dir)
  File "/nix/store/81lwy2hfqj4c1943b1x8a0qsivjhdhw9-python3-3.9.6/lib/python3.9/venv/__init__.py", line 116, in e
nsure_directories
    create_if_needed(env_dir)
  File "/nix/store/81lwy2hfqj4c1943b1x8a0qsivjhdhw9-python3-3.9.6/lib/python3.9/venv/__init__.py", line 105, in c
reate_if_needed
    os.makedirs(d)
  File "/nix/store/81lwy2hfqj4c1943b1x8a0qsivjhdhw9-python3-3.9.6/lib/python3.9/os.py", line 215, in makedirs
    makedirs(head, exist_ok=exist_ok)
  File "/nix/store/81lwy2hfqj4c1943b1x8a0qsivjhdhw9-python3-3.9.6/lib/python3.9/os.py", line 215, in makedirs
    makedirs(head, exist_ok=exist_ok)
  File "/nix/store/81lwy2hfqj4c1943b1x8a0qsivjhdhw9-python3-3.9.6/lib/python3.9/os.py", line 215, in makedirs
    makedirs(head, exist_ok=exist_ok)
  File "/nix/store/81lwy2hfqj4c1943b1x8a0qsivjhdhw9-python3-3.9.6/lib/python3.9/os.py", line 225, in makedirs
    mkdir(name, mode)
OSError: [Errno 30] Read-only file system: '/nix/store/p3j4334jxdrdwfqj02zy605jc5k8h16z-vimplugin-chadtree-2021-0
8-10/share/vim-plugins/chadtree/.vars'
Press ENTER or type command to continue
```

**Comment by ms-jpq** [comment]
https://github.com/ms-jpq/chadtree/pull/222

I think I fixed it, sorry, will do this for `coq` soon

**Comment by cideM** [comment]
@arcticlimer I'm the person who created the original issue you linked to and I was thinking about how to package this plugin properly just now.

There's a very nice project https://github.com/DavHau/mach-nix which makes it easy to generate Nix expressions from and for Python packages and applications. I think that it should be possible to use this to properly package Coq. By properly I mean that the dependencies are managed by Nix, rather than downloading them manually through the plugin itself, which isn't very Nix-y.

I also know that you can already add some packages to the Python interpreter used by Neovim in Nix: https://discourse.nixos.org/t/adding-python-packages-to-the-python-interpreter-inside-the-neovim-derivation/9507

I don't feel like putting all of these things together right now, because it's probably still quite a bit of work, but maybe you have a bit more motivation. In any case, I'll just leave this information here in case someone else stumbles upon this thread.

**Comment by ms-jpq** [comment]
thing is right, with like `nix` being basically still quite niche, I don't want to dedicate too much effort into making an actual `nix` package, because functionally, downloading everything version locked through `pip` and doing clean installs works quite well across basically windows, linux, mac, and even bsd.

`coq` already wipes `venv` if the lock files change, so everything is both local and reproducible.

I will add in support for using `coq` under readonly constraints and use `nvim`'s own `stdpath` for storage (which is XDG compliant on *nix, and under `<username>/AppData/` for windows), because stuff like `ctags` need to go somewhere.

so basically `coq` will be `nix` compatible, but not have it packaged specifically for `nix`.

**Comment by cideM** [comment]
Totally understandable! I wasn't implying that I expect you to package your app for a niche OS you don't even use :pray: ! I really appreciate you making the changes that will allow Nix users to use the plugin in a read-only environment. Since I have some free time on my hands, I'll try packaging it myself actually ^_^

Started something here https://discourse.nixos.org/t/how-to-package-neovim-plugin-with-python-application/14683

**Comment by ms-jpq** [comment]
coming next PR https://github.com/ms-jpq/coq_nvim/pull/165

**Comment by ms-jpq** [comment]
pr merged
